src/item.cpp

```diff
@@ -3829,13 +3803,32 @@ std::string item::gun_type() const
 
 skill_id item::weap_skill() const
 {
-    if( !is_weap() && !is_tool() ) {
+    if( is_null() || is_gun() || is_gunmod() || is_food() || is_ammo() ||
```

Comment:
> If you're using a gun with a bayonet as a spear, it should train spears (and vice versa).

That would mean both `is_gun()` and `is_melee()` could be true at the same time. That isn't necessarily a problem but we would need that so code continues to perform the checks in that order (currently the case for all existing code)

> The null item shouldn't have a skill
> > It should - it is treated as the fist in pretty much all cases.

I'm less sure about this. The only cases where it matters were the two functions above